### PR TITLE
Fix #2358: Test meta refresh parsing

### LINKED ISSUE #2358: Import/convert meta refresh tests
See
https://github.com/whatwg/html/pull/246#issuecomment-149152666
and
https://github.com/whatwg/html/pull/246#issuecomment-149477448

### PR DESCRIPTION
**Pull request by zcorpan**
Add tests for https://github.com/whatwg/html/pull/246.

This needs https://github.com/w3c/wptserve/pull/68 to function correctly.

cc @jgraham

### CODE REVIEW

**Comment by zcorpan** [comment]
@domenic since you reviewed the spec change and have looked at those tests already, want to review this also?

File: html/semantics/document-metadata/the-meta-element/pragma-directives/attr-meta-http-equiv-refresh/parsing.html [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  {input: '1; urlfoo', expected: [1, 'urlfoo']},
+  {input: '1; urfoo', expected: [1, 'urfoo']},
+  {input: '1; ufoo', expected: [1, 'ufoo']},
+  {input: '1; "foo"bar', expected: [1, 'foo']},
+  {input: '; foo', expected: []},
+  {input: ', foo', expected: []},
+  {input: 'foo', expected: []},
+  {input: '+1; url=foo', expected: []},
+  {input: '-1; url=foo', expected: []},
+  {input: '-0; url=foo', expected: []},
```
Comment (domenic) [inline html/semantics/document-metadata/the-meta-element/pragma-directives/attr-meta-http-equiv-refresh/parsing.html]: Maybe also test +0 and 0?

File: html/semantics/document-metadata/the-meta-element/pragma-directives/attr-meta-http-equiv-refresh/parsing.html [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  {input: '1; url="foo\'bar', expected: [1, 'foo\'bar']},
+  {input: '1; url foo', expected: [1, 'url foo']},
+  {input: '1; urlfoo', expected: [1, 'urlfoo']},
+  {input: '1; urfoo', expected: [1, 'urfoo']},
+  {input: '1; ufoo', expected: [1, 'ufoo']},
+  {input: '1; "foo"bar', expected: [1, 'foo']},
+  {input: '; foo', expected: []},
+  {input: ', foo', expected: []},
+  {input: 'foo', expected: []},
+  {input: '+1; url=foo', expected: []},
```
Comment (domenic) [inline html/semantics/document-metadata/the-meta-element/pragma-directives/attr-meta-http-equiv-refresh/parsing.html]: Maybe also test these weird number cases without url=?

File: html/semantics/document-metadata/the-meta-element/pragma-directives/attr-meta-http-equiv-refresh/parsing.html [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+// failure to parse is []
+// success to parse is [time, url] where url is unresolved
+
+var tests_arr = [
+  {input: '', expected: []},
+  {input: '1', expected: [1, 'refresh.sub.html']},
+  {input: '1;', expected: [1, 'refresh.sub.html']},
+  {input: '1,', expected: [1, 'refresh.sub.html']},
+  {input: '1; url=foo', expected: [1, 'foo']},
```
Comment (domenic) [inline html/semantics/document-metadata/the-meta-element/pragma-directives/attr-meta-http-equiv-refresh/parsing.html]: Maybe test more whitespace in between the punctuation and the url= (or lack of url=)?

**Comment by zcorpan** [comment]
@domenic thx, comments addressed

File: html/semantics/document-metadata/the-meta-element/pragma-directives/attr-meta-http-equiv-refresh/parsing.html [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  {input: '+1; url=foo', expected: []},
+  {input: '-1; url=foo', expected: []},
+  {input: '+0; url=foo', expected: []},
+  {input: '-0; url=foo', expected: []},
+  {input: '0; url=foo', expected: [0, 'foo']},
+  {input: '+1', expected: []},
+  {input: '-1', expected: []},
+  {input: '+0', expected: []},
+  {input: '-0', expected: []},
+  {input: '0', expected: [0, 'refresh.sub.html']},
```
Comment (domenic) [inline html/semantics/document-metadata/the-meta-element/pragma-directives/attr-meta-http-equiv-refresh/parsing.html]: What about `0; foo`, `+1; foo`, etc.?

File: html/semantics/document-metadata/the-meta-element/pragma-directives/attr-meta-http-equiv-refresh/parsing.html [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  {input: '1\t', expected: [1, 'refresh.sub.html']},
+  {input: '1\r', expected: [1, 'refresh.sub.html']},
+  {input: '1\n', expected: [1, 'refresh.sub.html']},
+  {input: '1\f', expected: [1, 'refresh.sub.html']},
+  {input: '1;', expected: [1, 'refresh.sub.html']},
+  {input: '1,', expected: [1, 'refresh.sub.html']},
+  {input: '1; url=foo', expected: [1, 'foo']},
+  {input: '1, url=foo', expected: [1, 'foo']},
+  {input: '1 url=foo', expected: [1, 'foo']},
+  {input: '1;\turl=foo', expected: [1, 'foo']},
```
Comment (domenic) [inline html/semantics/document-metadata/the-meta-element/pragma-directives/attr-meta-http-equiv-refresh/parsing.html]: Oooh, nice, I didn't think to test \t and friends. What about multiple spaces, e.g. `1;      url=foo`? Or `1,url      = foo` or `1 url    =    foo`?

**Comment by zcorpan** [comment]
@domenic ok

```diff
@@ -41,7 +41,11 @@ var tests_arr = [
   {input: '1x;url=foo', expected: []},
   {input: '1 x;url=foo', expected: [1, 'x;url=foo']},
   {input: '1;;url=foo', expected: [1, ';url=foo']},
-  {input: ' 1 ; url = foo', expected: [1, 'foo']},
+  {input: '  1  ;  url  =  foo', expected: [1, 'foo']},
+  {input: '  1  ,  url  =  foo', expected: [1, 'foo']},
+  {input: '  1  ;  foo', expected: [1, 'foo']},
+  {input: '  1  ,  foo', expected: [1, 'foo']},
+  {input: '  1  url  =  foo', expected: [1, 'foo']},
   {input: '1; url=foo ', expected: [1, 'foo']},
   {input: '1; url=f\to\no', expected: [1, 'foo']},
   {input: '1; url="foo"bar', expected: [1, 'foo']},
@@ -60,6 +64,11 @@ var tests_arr = [
   {input: '+0; url=foo', expected: []},
   {input: '-0; url=foo', expected: []},
   {input: '0; url=foo', expected: [0, 'foo']},
+  {input: '+1; foo', expected: []},
+  {input: '-1; foo', expected: []},
+  {input: '+0; foo', expected: []},
+  {input: '-0; foo', expected: []},
+  {input: '0; foo', expected: [0, 'foo']},
   {input: '+1', expected: []},
   {input: '-1', expected: []},
   {input: '+0', expected: []},
```

**Comment by domenic** [comment]
LGTM! Curious to see the browser pass matrix on this!!

**Comment by domenic** [comment]
Hmm, I should have run these on Chrome before LGTMing... https://w3c-test.org/html/semantics/document-metadata/the-meta-element/pragma-directives/attr-meta-http-equiv-refresh/parsing.html shows 63 timeouts, 1 success :(.

Full stats:

- Firefox 46 pass/16 fail/2 timeout
- Chrome 1 pass/0 fail/63 timeout
- Edge 25 pass/39 fail/0 timeout

Edge also has a lot of "permission denied" errors.

Curious what _new_ interop problems we've discovered here.

**Comment by zcorpan** [comment]
Maybe this test needs a long timeout when run non-locally. And maybe I should create "foo" etc files so it doesn't check `location` on a 404.

**Comment by zcorpan** [comment]
Locally in chrome canary I get
46 Pass
10 Fail
8 Timeout

**Comment by domenic** [comment]
LGTM; tested locally this time.

If this still doesn't help we may want to investigate using promise_test instead of async_test in order to get the tests to run sequentially instead of in parallel.

**Comment by zcorpan** [comment]
Did you mean to lgtm #2377?

**Comment by domenic** [comment]
Er, yes.

**Pull request merged**